src/HotChocolate/Core/src/Execution/ExecutionContext.cs

```diff
@@ -0,0 +1,100 @@
+using System.Threading;
+using System.Threading.Tasks;
+using HotChocolate.Fetching;
+using HotChocolate.Execution.Utilities;
+using Microsoft.Extensions.ObjectPool;
+using System;
+
+namespace HotChocolate.Execution
+{
+    internal class ExecutionContext : IExecutionContext
+    {
+        private readonly TaskQueue _taskQueue;
+        private readonly TaskStatistics _taskStatistics;
+        private readonly object _engineLock = new object();
+        private TaskCompletionSource<bool>? _waitForEngineTask;
+
+        public ExecutionContext(
+            ObjectPool<ResolverTask> taskPool,
+            IBatchDispatcher batchDispatcher)
+        {
+            _taskStatistics = new TaskStatistics();
+            _taskQueue = new TaskQueue(_taskStatistics, taskPool);
+            TaskPool = taskPool;
+            BatchDispatcher = batchDispatcher;
+            BatchDispatcher.TaskEnqueued += BatchDispatcherEventHandler;
+            TaskStats.StateChanged += TaskStatisticsEventHandler;
+        }
+
+        public ITaskQueue Tasks => _taskQueue;
+
+        public ITaskStatistics TaskStats => _taskStatistics;
+
+        public bool IsCompleted => TaskStats.Enqueued == 0 && TaskStats.Running == 0;
+
+        public ObjectPool<ResolverTask> TaskPool { get; }
+
+        public IBatchDispatcher BatchDispatcher { get; }
+
+        public Task WaitForEngine(CancellationToken cancellationToken)
+        {
+            lock (_engineLock)
```

Comment:
I think we could get rid of this lock if we had a local variable that wirst adds the wait for engine task.

```csharp
TaskCompletionSource<bool>? waitForEngineTask = _waitForEngineTask; 
if (waitForEngineTask is null)
{
    return Task.CompletedTask;
}

cancellationToken.Register(() => waitForEngineTask.SetCanceled());
return waitForEngineTask.Task;
```